### PR TITLE
Implement generators

### PR DESCRIPTION
**Pull request by Aaron1011**
`GeneratorDatum` and `GeneratorWitnessDatum` are introduced, which
mirror `TyKind::Generator` and `TyKind::GeneratorWitness` in rustc.

We handle auto traits for `GeneratorDatum` using the existing
`constituent_types` method. For `GeneratorWitnessDatum`, we generate a
`forall<'a, 'b, ... 'z>` goal to reflect the fact that our lifetimes are
existentially bound.

Unresolved questions:
* The implementation of `GeneratorWitnessDatum` needs careful review -
  I'm not certain that I fully understand how `Binders` works.
* The syntax implemented in `parser.lalrpop` can be bikesheeded. We
  might want to use something other than `for<>` for the witness types,
  to reflect that fact that we have existential rather than universal
  quantification.
* The generator grammar only allows quantifying over liftimes for the
  witness types - however, this gets lowered to a `Binders`, which
  supports types and constants as well. Should we do anything else
  to ensure we never end up with types/consts where they're not
  expected?
* I added a test to ensure that we treat the witness lifetimes as
  existentially quantified - for example, if we have
  `for<'a, 'b> Foo<'a, 'b>`, we should only be able to use an auto
  trait impl that is fully generic over lifetimes - e.g.
  `impl<'a> Send for Foo<'a, 'b>` will not be used. I *believe*
  that the test is showing this behavior - it ends up returning
  region constraints that require `'a` and `'b` to outlive each other.
  Since these are 'existential' lifetimes (which cannot be substituted),
  this is an unsatisfiable constraint. However, I'm not sure where we
  should be detecting that this is unsatisfiable.
* In rustc, all of the existential witness lifetimes are distinct. This is not enforced by this PR e.g. you can repeat a lifetime in the witness type. I'm not sure if/where we should enforce this.

### CODE REVIEW

**Comment by Aaron1011** [comment]
This is mostly working - however, I still need to figure out how to treat generator witness lifetime as existentially quantified when handling auto traits.

**Comment by Aaron1011** [comment]
@jackh726 @nikomatsakis This is ready for review

**Comment by jackh726** [comment]
I definitely don't know enough about generators to make a meaningful review of this, so will definitely have to defer to @nikomatsakis. However, the book builtin types table will need to be updated here.

Additionally, I really want to push for us to document generators and generator witnesses in the chalk book. I don't think it *needs* to be in this PR, but I am somewhat worried that if it isn't at least loosely documented here, it'll get pushed aside. Particularly, I feel like documenting this is super important since there's almost no documentation in the rustc dev guide or in the rustc codebase (that I can find).

**Comment by Aaron1011** [comment]
I'll work on writing up some docs.

**Comment by jackh726** [comment]
> I'll work on writing up some docs.

Awesome!

I do want to stress that I understand docs can sometimes be hard to write. But I would definitely error on the side of "some docs are better than no docs" and "don't let perfect be the enemy of good enough"; we can always expand upon/clarify the docs in subsequent PRs.

File: book/src/types/rust_types/application_ty.md [hunk 1]
```diff
@@ -22,5 +22,47 @@ _)`) or "fixed-length array" `[_; _]`. Note that the precise set of
...
+
+A `Generator` represents a Rust generator. There are three major components
+to a generator:
+
+* Upvars - similar to closure upvars, they reference values outside of the generator,
+  and are stored across al yield points.
+* Resume/yield/return types - the types produced/consumed by various generator methods.
+  These are not stored in the generator across yield points - they are only
+  used when the generator is running.
+* Generator witness - see the `Generator Witness` section below.
```
Comment (nikomatsakis) [inline book/src/types/rust_types/application_ty.md:37]: thank you for writing docs! wonderful.

File: book/src/types/rust_types/application_ty.md [hunk 2]
```diff
@@ -22,5 +22,47 @@ _)`) or "fixed-length array" `[_; _]`. Note that the precise set of
...
+
+Of these types, only upvars and resume/yield/return are stored directly in
+`TypeName::Generator`. The generator witness is implicitly associated with the generator
+by virtue of sharing the same `GeneratorId`. It is only used when determining auto trait
+impls, where it is considered a 'constituent type'.
+
+### Generator witness types
+
+The `GeneratorWitness` variant represents the generator witness of
+the generator with id `GeneratorId`.
```
Comment (nikomatsakis) [inline book/src/types/rust_types/application_ty.md:47]: I think it'd be good to explain the *purpose* of a witness:

The witness type captures the types of all variables that may be live across a `yield` point.

File: chalk-solve/src/clauses.rs [hunk 3]
```diff
@@ -212,6 +237,89 @@ pub fn push_auto_trait_impls_opaque<I: Interner>(
...
+        // hold for the original lifetime (before it was erased).
+        //
+        // This does not take into account well-formed information from
+        // the witness types. For example, if we have the type
+        // `struct Foo<'a, 'b> { val: &'a &'b u8 }`
+        // then `'b: 'a` must hold for `Foo<'a, 'b>` to be well-formed.
+        // If we have `Foo<'a, 'b>` stored as a witness type, we will
+        // not currently use this information to determine a more precise
+        // relationship between 'a and 'b. In the future, we will likely
+        // do this to avoid incorrectly rejecting correct code.
```
Comment (nikomatsakis) [inline chalk-solve/src/clauses.rs:298]: Another great comment.

File: tests/test/generators.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            }
+
+            generator upvar_lifetime_restrict<T>[resume = (), yield = ()] {
+                upvars [T; StructOne]
+                witnesses for<'a, 'b> [SendSameLifetime<'a, 'b, T>]
+            }
+
+            generator send_any_lifetime<T>[resume = (), yield = ()] {
+                upvars []
+                witnesses for<'a, 'b> [SendAnyLifetime<'a, 'b, T>; u8]
```
Comment (nikomatsakis) [inline tests/test/generators.rs:32]: the `for` syntax here seems a bit misleading, I guess?

**Review by nikomatsakis — APPROVED** [review]
This is great. I left a nit about the syntax but I'm inclined to land it.

File: tests/test/generators.rs (hunk 4 above)
Comment (Aaron1011) [reply tests/test/generators.rs:32]: Yeah, it's not great. Do you think `exists<>` would be better?

File: book/src/types/rust_types/application_ty.md (hunk 2 above)
Comment (Aaron1011) [reply book/src/types/rust_types/application_ty.md:47]: I just moved this entire block from a different section of the book :) - I'm happy to expand on it.

**Comment by Aaron1011** [comment]
@nikomatsakis: I've addressed your comments

**Pull request merged**